Sketch out gen module

justanotherdot opened the pull request:
Now that seed and size are 'ported', we can work on getting the
functionality from the gen module over here, too.

See #18.

justanotherdot committed:
Sketch out gen module

Now that seed and size are 'ported', we can work on getting the
functionality from the gen module over here, too.

justanotherdot committed:
Drop unused imports

justanotherdot committed:
From and to random

justanotherdot committed:
delay impl + Random type synonym

justanotherdot committed:
Broken start to unfold and unfold_forest

justanotherdot committed:
Fix unfold but break create

justanotherdot committed:
Lifetime woes with `create`

justanotherdot committed:
More tweaks to unfold and friends

justanotherdot committed:
Possibly non-working impl of create

justanotherdot committed:
When in doubt, reference count

justanotherdot committed:
Accept boxed fns for create

justanotherdot committed:
Finalise create impl

justanotherdot committed:
PartialEq impl for Tree + non-mut forcing thunks

justanotherdot committed:
Tidy up Lazy test

justanotherdot committed:
Teach thunks to actually memoize values

justanotherdot committed:
Comment on Lazy use of RefCell

justanotherdot committed:
gen::shrink et. al.

justanotherdot committed:
gen::sized et. al.

justanotherdot committed:
gen::resize et. al.

justanotherdot committed:
gen::scale et. al.

justanotherdot commented:
While shuffling around the furniture I noticed the `Lazy` struct isn't really memoizing values so I've fixed that here while hashing out porting stuff 1:1 from (mostly) the F# port. I'll also be changing the ergonomics s.t. `Lazy` will just accept generic values and wrap them in a closure themselves (rather than relying on the client to do that which is pointless).

justanotherdot commented:
Actually, one case for passing along a function to `Lazy` is to allow it to suspend values that would otherwise be evaluated eagerly. I'll see if I can provide both although I suspect that latter case is probably not so common.

justanotherdot committed:
Simplify ctor for Lazy

justanotherdot committed:
gen::integral et. al.

justanotherdot committed:
Bind impls for gen, random, and tree

justanotherdot committed:
gen::item et. al.

justanotherdot committed:
Use direct gen::constant in gen::item

justanotherdot commented:
This is coming along and getting pretty large-ish for my liking so I might stop here and do the rest of the gen module in another PR. @jacobstanley do you mind taking a look through for anything that seems fishy? 🐠

justanotherdot commented on src/seed.rs:
```diff
@@ -62,22 +62,22 @@ pub fn next_double(lo: f64, hi: f64, mut s0: Seed) -> (f64, Seed) {
 pub fn mix64_variant13(x: u64) -> u64 {
-    let y = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9;
-    let z = (y ^ (y >> 27)) * 0x94d049bb133111eb;
+    let y = (x ^ (x >> 30)).wrapping_mul(0xbf58476d1ce4e5b9);
```
I had to do these as I was getting panics from rust when running tests that indirectly called `seed` fns.

jacobstanley replied:
really? what does this do?

justanotherdot replied:
so rust will panic by default on overflow/underflow from ops, this just goes back to the default you'd expect in other languages

justanotherdot commented on src/tree.rs:
```diff
@@ -1,22 +1,137 @@
     }
 
-    pub fn value(&mut self) -> &A {
+    pub fn value(&self) -> Option<A> {
```
this isn't great as we always expect some value of A to come back, but it means a fair bit of play that's a tad distracting at the moment. I'll make an issue and come back to it later.

justanotherdot committed:
Fix curried function errors

justanotherdot commented:
This is getting a bit unwieldy. Would love some eyes for quick sanity and then happy to merge and move on where I left off (which will be on `choice` from [the F# port](https://github.com/hedgehogqa/fsharp-hedgehog/blob/master/src/Hedgehog/Gen.fs#L238-L248))

justanotherdot commented on src/lazy.rs:
```diff
@@ -1,49 +1,81 @@
+where
+    A: Clone + 'a,
+{
+    // TODO: We can simplify the ergonomics of this interface by not requiring a closure.
```
dead comment

justanotherdot commented on src/gen.rs:
```diff
@@ -1,18 +1,238 @@
+    }
+}
+
+//pub fn frequency<'a, I, A>(xs0: I) -> Gen<'a, A>
```
will tackle this in followup PR

jacobstanley replied:
not a high priority i wouldn't think

justanotherdot commented on src/lazy.rs:
```diff
@@ -1,49 +1,81 @@
+        }
+    }
+
+    // TODO: We can simplify the ergonomics of this interface by not requiring a closure.
```
dead comment

jacobstanley commented on src/gen.rs:
```diff
@@ -1,18 +1,238 @@
+}
+
+pub fn to_random<'a, A>(g: Gen<'a, A>) -> Random<Tree<'a, A>> {
+    g.0
```
this is rust for unpack newtype?

justanotherdot replied:
Yeah they have the same form as a 1-tuple

justanotherdot commented on src/shrink.rs:
```diff
@@ -52,13 +52,14 @@ where
 
 /// Shrink an integral number by edging towards a destination.
-pub fn towards<'a, A: 'a>(destination: A) -> impl Fn(A) -> Vec<A>
+pub fn towards<'a, 'b, A>(destination: A) -> impl Fn(A) -> Vec<A>
```
this lifetime might not be needed...

justanotherdot commented on src/tree.rs:
```diff
@@ -1,22 +1,138 @@
+        let has_children_str = if self.children.is_empty() {
+            format!("<>")
+        } else {
+            format!("<children>")
```
being lazy for the moment, although theoretically this is a simple implementation; you just go through each child and call fmt on them.

jacobstanley replied:
this can't be derived automatically?

justanotherdot replied:
Fixed in 4a8e554

jacobstanley commented on src/range.rs:
```diff
@@ -258,7 +260,7 @@ where
     Range(
         z.clone(),
-        Box::new(move |sz| {
+        Rc::new(move |sz| {
```
so what was the deal here? all the Box -> Rc

justanotherdot replied:
A `Box` can only be cloned if the thing it's boxing is cloneable. When you clone an Rc you are simply incrementing the count. This is so I can actually hand functions around to things on different levels since I can't clone functions.

justanotherdot commented:
No, because a tree embeds a lazy which embeds a function which doesn’t have
a debug instance

On Fri, 14 Jun 2019 at 10:15 am Jacob Stanley <notifications@github.com>
wrote:

> *@jacobstanley* commented on this pull request.
> ------------------------------
>
> In src/tree.rs
> <https://github.com/justanotherdot/rust-hedgehog/pull/22#discussion_r293622116>
> :
>
> > +            thunk: t1.thunk,
> +            children: xs,
> +        }
> +    }
> +}
> +
> +impl<'a, A> Debug for Tree<'a, A>
> +where
> +    A: 'a + Clone + Debug,
> +{
> +    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
> +        // TODO: Actually print the children.
> +        let has_children_str = if self.children.is_empty() {
> +            format!("<>")
> +        } else {
> +            format!("<children>")
>
> this can't be derived automatically?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/justanotherdot/rust-hedgehog/pull/22?email_source=notifications&email_token=ABI2FEALFFHZWLEX7FWCGQTP2LPLDA5CNFSM4HKHNQDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3QYYXA#discussion_r293622116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABI2FEEI36KOWO2E5GLYCZLP2LPLDANCNFSM4HKHNQDA>
> .
>

jacobstanley commented:
presumably you can't write a debug instance for Lazy?

justanotherdot commented:
ah right! yes, I could do that and just `<Function>` out the function as a stub. I guess my question is: should showing something via `Debug` force the thunk? Seems like I'm already doing this with the Tree impl, anyways.

jacobstanley commented:
I think yes (if you're allowed to do it anyway, even unsafely)

justanotherdot commented:
Nice! I'll put that in sometime today 👍

jacobstanley commented:
lol @amosr and i used to use `Show` to implement `DeepSeq` as a quick hack

justanotherdot committed:
Derive Debug for Tree

This is through specifying an impl for Debug on Lazy.

This now makes the constraint for `A` being `Clone` much more strict
(thus the relevant changes). I'm surprised this didn't need enforcing
before.

justanotherdot commented:
Alright, I've updated the `Debug` instance on `Tree` as a automatic derivation and specified a manual implementation for `Debug` on `Lazy`.

justanotherdot committed:
Remove dead lifetime

justanotherdot committed:
Rm dead comments

justanotherdot commented:
bors delegate=[jacobstanley]

justanotherdot commented:
bors r+
